### ISSUE TITLE
SciRegressor Performance Degradation with Scaling - R2 Calculation Issues During    Hyperparameter Tuning

### ISSUE THREAD

**Issue by sandrohuni** [hydrosolutions/long-term-forecasting#4]
⏺ SciRegressor Performance Degradation with Scaling - R2 Calculation Issues During 
  Hyperparameter Tuning

  Problem Description

  When enabling scaling/normalization in SciRegressor, hyperparameter tuning reports
  very poor R2 values, despite the model potentially performing well. This is
  particularly problematic for hydrological forecasting where past discharge values
  are used as features to predict future discharge - a relationship that requires
  consistent scaling across features and targets.

  Context: Why Both Features and Targets Must Be Scaled

  In monthly discharge forecasting:
  - Past discharge values (t-1, t-2, etc.) are used as features
  - Future discharge (t+1) is the target
  - These share the same scale and physical meaning
  - Scaling only features while leaving targets unscaled would destroy the temporal
  autocorrelation structure

  Therefore, the current implementation correctly scales both features and targets
  together.

  Root Cause

  The issue is not with the scaling approach itself, but with how R2 is calculated 
  during hyperparameter tuning. The code calculates R2 on scaled values, which can be
   misleading because:

  1. R2 measures explained variance
  2. Scaling changes the variance structure
  3. A good model in scaled space might show poor R2 if the scaling dramatically
  reduces variance

  Current Behavior

  # In tune_hyperparameters() - PROBLEMATIC
  Y_val_scaled = scale_values(Y_val)
  predictions_scaled = model.predict(X_val_scaled)
  self.best_r2 = r2_score(Y_val_scaled, predictions_scaled)  # R2 on scaled values!

  This calculates R2 in the scaled space, which may not reflect true model
  performance.

  Technical Analysis

  Why R2 Appears Poor with Scaling:

  1. Variance Reduction: If discharge has high variance (e.g., std=1000), scaling to
  std=1 reduces variance by 10^6
  2. R2 Sensitivity: Small prediction errors in scaled space can yield poor R2 even
  if unscaled predictions are good
  3. Scale Mismatch: Different basins might have vastly different discharge scales,
  making global scaling problematic

  Example Scenario:
  - Original discharge range: 0-5000 m³/s
  - Scaled discharge range: -2 to +2 (standardized)
  - A 0.1 unit error in scaled space = 125 m³/s in original space
  - This can drastically affect R2 in scaled vs original space

  Proposed Solution

  Calculate R2 in the original scale for meaningful performance metrics:

  # In tune_hyperparameters() - CORRECTED
  def tune_hyperparameters(self, ...):
      # ... existing code ...

      # Train on scaled data (correct)
      model.fit(X_train_scaled, Y_train_scaled)
      predictions_scaled = model.predict(X_val_scaled)

      # Inverse transform for R2 calculation
      predictions_original = inverse_transform(predictions_scaled,
                                            mean=discharge_mean,
                                            std=discharge_std)
      Y_val_original = inverse_transform(Y_val_scaled,
                                       mean=discharge_mean,
                                       std=discharge_std)

      # Calculate R2 in original scale
      self.best_r2 = r2_score(Y_val_original, predictions_original)

  Implementation Requirements

  1. Store scaling parameters during data processing:
  self.discharge_scaler_params = {
      'mean': discharge_mean,
      'std': discharge_std,
      'type': normalization_type
  }

  2. Implement inverse transform for discharge:
  def inverse_transform_discharge(self, scaled_values):
      if self.discharge_scaler_params['type'] == 'global':
          return scaled_values * self.discharge_scaler_params['std'] + \
                 self.discharge_scaler_params['mean']
      # Handle other normalization types...

  3. Ensure consistent scaling between features and targets in the discharge columns

  Alternative Considerations

  1. Report Multiple Metrics: Show both scaled and unscaled R2 for transparency
  2. Use RMSE: Consider RMSE in original units as primary metric
  3. Basin-Specific Evaluation: Evaluate per-basin performance when using per-basin
  scaling

  Acceptance Criteria

  1. ✅ R2 during hyperparameter tuning reflects actual model performance
  2. ✅ Scaling parameters are properly stored and accessible for inverse
  transformation
  3. ✅ Both features and targets maintain consistent scaling (especially discharge
  columns)
  4. ✅ Model selection based on meaningful performance metrics
  5. ✅ Clear documentation on the scaling strategy and its implications

  Testing Strategy

  1. Synthetic Test: Create data where scaled and unscaled R2 should differ
  significantly
  2. Comparison Test: Train same model with/without scaling, verify similar actual
  performance
  3. Inverse Transform Test: Verify predictions can be accurately transformed back
  4. Multi-Basin Test: Ensure per-basin scaling works correctly with inverse
  transform

  Priority: High

  This issue causes incorrect model selection during hyperparameter tuning,
  potentially choosing suboptimal models.

  Notes

  - The current approach of scaling both features and targets is correct for
  discharge forecasting
  - The issue is purely about performance measurement, not the scaling strategy
  - This explains why models perform poorly during tuning but might work well in
  practice
